onnxruntime/core/providers/coreml/builders/impl/LRN_op_builder.cc

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/coreml/builders/helper.h"
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class LRNOpBuilder : public BaseOpBuilder {
+  // Add operator related
+#ifdef __APPLE__
+ private:
+  [[nodiscard]] Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                             const logging::Logger& logger) const override;
+#endif
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+
+#ifdef __APPLE__
+
+Status LRNOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                           const Node& node,
+                                           const logging::Logger& logger) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+
+  auto* coreml_lrn = layer->mutable_lrn();
+
+  NodeAttrHelper helper(node);
+  const auto alpha = helper.Get("alpha", 0.0001f);
+  const auto beta = helper.Get("beta", 0.75f);
+  const auto bias = helper.Get("bias", 1.0f);  // k
+  const auto size = helper.Get("size", 1);     // localSize
+
+  coreml_lrn->set_alpha(alpha);
+  coreml_lrn->set_beta(beta);
+  coreml_lrn->set_localsize(size);
+  coreml_lrn->set_k(bias);
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+  return Status::OK();
+}
+#endif
+
+// Operator support related
+
+bool LRNOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& /*input_params*/,
+                                     const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "LRN does not support empty input shape";
+    return false;
+  }
+
+  const auto input_rank = input_shape.size();
+  if (input_rank < 3) {
```

Comment:
from the [CoreML LRN spec](https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#lrnlayerparams):

"For ranks greater than 3, the leading dimensions, starting from 0 to -4 (inclusive), are all treated as batch."

I don't think [ONNX LRN](https://github.com/onnx/onnx/blob/main/docs/Operators.md#LRN) also does this. should we limit it to supporting input rank 3 or 4?

Reply:
okay I see. will update.

Reply:
Good catch.